test/development/app-hmr/hmr.test.ts

```diff
@@ -73,70 +70,77 @@ describe(`app-dir-hmr`, () => {
+                'window.__TEST_NO_RELOAD === undefined'
+              )
+              // Used to be flaky but presumably no longer is.
+              // If this flakes again, please add the received value as a commnet.
+              expect({ envValue, mpa }).toEqual({
+                envValue: 'ipad',
+                mpa: false,
               })
-              // FIXME: Should be either a single "rebuilding"+"done" or the last "rebuilding" should be followed by "done"
-              expect(fastRefreshLogs).toEqual([
-                { source: 'log', message: '[Fast Refresh] rebuilding' },
-                { source: 'log', message: '[Fast Refresh] rebuilding' },
-                {
-                  source: 'log',
-                  message: expect.stringContaining('[Fast Refresh] done in '),
-                },
-                { source: 'log', message: '[Fast Refresh] rebuilding' },
-              ])
             },
             // Very slow Hot Update for some reason.
             // May be related to receiving 3 rebuild events but only one finish event
             5000
           )
+
+          const fastRefreshLogs = logs.filter((log) => {
+            return log.message.startsWith('[Fast Refresh]')
+          })
+          expect(fastRefreshLogs).toEqual([
+            { source: 'log', message: '[Fast Refresh] rebuilding' },
+            {
+              source: 'log',
+              message: expect.stringContaining('[Fast Refresh] done in '),
+            },
+            { source: 'log', message: '[Fast Refresh] rebuilding' },
+            { source: 'log', message: '[Fast Refresh] rebuilding' },
+            {
+              source: 'log',
+              message: expect.stringContaining('[Fast Refresh] done in '),
+            },
+            {
+              source: 'log',
+              message: expect.stringContaining('[Fast Refresh] done in '),
+            },
+          ])
         }
-        const envValue = await browser.elementByCss('p').text()
-        const mpa = await browser.eval('window.__TEST_NO_RELOAD === undefined')
-        // Flaky sometimes in Webpack:
-        // A. misses update and just receives `{ envValue: 'mac', mpa: false }`
-        // B. triggers error on server resulting in MPA: `{ envValue: 'ipad', mpa: true }` and server logs: ⨯ [TypeError: Cannot read properties of undefined (reading 'polyfillFiles')] ⨯ [TypeError: Cannot read properties of null (reading 'default')]
-        // A is more common than B.
-        expect({ envValue, mpa }).toEqual({
-          envValue:
-            isPPREnabledByDefault && !process.env.TURBOPACK
-              ? // FIXME: Should be 'ipad' but PPR+Webpack swallows the update reliably
-                'mac'
-              : 'ipad',
-          mpa: false,
-        })
       } finally {
-        await next.patchFile(envFile, envContent)
+        // TOOD: use sandbox instead
+        await next.patchFile(envFile, 'MY_DEVICE="mac"')
```

Comment:
Don't we already have `sandbox` for that?

Reply:
`sandbox` stops and starts the server. The third argument of `patchFile` is useful for when you want to edit a file while the server is running, e.g. for HMR, without needing to manually cleanup the changes afterwards, so that they don't affect other tests.

Reply:
Huh, that's slightly confusing. If `sandbox` takes initial files that it patches, why doesn't it revert that patch on `cleanup()`?

Reply:
Yeah, that was confusingly phrased. Both are doing the cleanup, one implicitly, the other by calling `cleanup`. The difference is only whether you want to use the already running server without restarting.

Reply:
I also looked at the code and `cleanup()` from `sandbox` does not clean the initial files it was passed. It only calls `next.clean()` so if the `initialFiles` passed to `sandbox` are not the `initialFiles` passed to the `next` instance, it doesn't cleanup the patch.

I personally don't like the programmatic `initialFiles` since it doesn't allow to just run the test fixture without starting the test. So it seems to me we should get rid of `initialFiles` and make sure `sandbox` reverts patched files.